src/meta/src/barrier/worker.rs

```diff
@@ -311,19 +313,83 @@ impl<C: GlobalBarrierWorkerContext> GlobalBarrierWorker<C> {
                         }
                     }
                 },
-                (worker_id, resp_result) = self.control_stream_manager.next_collect_barrier_response() => {
-                    if let Err(e) = resp_result.and_then(|resp| self.checkpoint_control.barrier_collected(resp, &mut self.control_stream_manager)) {
-                        {
-
+                event = self.checkpoint_control.next_event() => {
+                    let result: MetaResult<()> = try {
+                        match event {
+                            CheckpointControlEvent::EnteringInitializing(entering_initializing) => {
+                                let database_id = entering_initializing.database_id();
+                                let error = merge_node_rpc_errors(&format!("database {} reset", database_id), entering_initializing.action.0.iter().filter_map(|(worker_id, resp)| {
+                                    resp.root_err.as_ref().map(|root_err| {
+                                        (*worker_id, ScoredError {
+                                            error: Status::internal(&root_err.err_msg),
+                                            score: Score(root_err.score)
+                                        })
+                                    })
+                                }));
+                                Self::report_collect_failure(&self.env, &error);
+                                if let Some(runtime_info) = self.context.reload_database_runtime_info(database_id).await? {
+                                    runtime_info.validate(database_id, &self.active_streaming_nodes).inspect_err(|e| {
+                                        warn!(database_id = database_id.database_id, err = ?e.as_report(), ?runtime_info, "reloaded database runtime info failed to validate");
+                                    })?;
+                                    entering_initializing.enter(runtime_info, &mut self.control_stream_manager);
+                                } else {
+                                    info!(database_id = database_id.database_id, "database removed after reloading empty runtime info");
+                                    entering_initializing.remove();
```

Comment:
The database item will not be removed, because there is no such command as `DropDatabase`, and we cannot tell whether the database is dropped totally, or just all streamings jobs are dropped but the database is still there. But in either case, the database will be an empty item in the `CheckpointControl`, and for this empty item, we won't do anything (even inject barrier) on it until a new streaming job is created on it.